File: sdks/ios/Sources/LegacyBaaSSDK/Models/Swift/MT103Message.swift
import Foundation

/// MT103 Single Customer Credit Transfer message
public struct MT103Message: Codable {
    /// Transaction reference number
    public let transactionReference: String
    
    /// Value date
    public let valueDate: Date?
    
    /// Currency code (ISO 4217)
    public let currency: String
    
    /// Transaction amount
    public let amount: Decimal
    
    /// Ordering customer information
    public let orderingCustomer: Customer
    
    /// Beneficiary customer information
    public let beneficiaryCustomer: Customer
    
    /// Ordering institution
    public let orderingInstitution: Institution?
    
    /// Beneficiary institution
    public let beneficiaryInstitution: Institution?
    
    /// Intermediary institution
    public let intermediaryInstitution: Institution?
    
    /// Remittance information
    public let remittanceInformation: String?
    
    /// Regulatory reporting information
    public let regulatoryReporting: String?
    
    /// Charges indication
    public let charges: ChargesIndication?
    
    public init(
        transactionReference: String,
        valueDate: Date? = nil,
        currency: String,
        amount: Decimal,
        orderingCustomer: Customer,
        beneficiaryCustomer: Customer,
        orderingInstitution: Institution? = nil,
        beneficiaryInstitution: Institution? = nil,
        intermediaryInstitution: Institution? = nil,
        remittanceInformation: String? = nil,
        regulatoryReporting: String? = nil,
        charges: ChargesIndication? = nil
    ) {
        self.transactionReference = transactionReference
        self.valueDate = valueDate
        self.currency = currency
        self.amount = amount
        self.orderingCustomer = orderingCustomer
        self.beneficiaryCustomer = beneficiaryCustomer
        self.orderingInstitution = orderingInstitution
        self.beneficiaryInstitution = beneficiaryInstitution
        self.intermediaryInstitution = intermediaryInstitution
        self.remittanceInformation = remittanceInformation
        self.regulatoryReporting = regulatoryReporting
        self.charges = charges
    }
}

/// Customer information for SWIFT messages
public struct Customer: Codable {
    /// Account number
    public let account: String
    
    /// Customer name
    public let name: String
    
    /// Customer address
    public let address: String?
    
    /// Country code
    public let country: String?
    
    /// Additional identification
    public let identification: String?
    
    public init(
        account: String,
        name: String,
        address: String? = nil,
        country: String? = nil,
        identification: String? = nil
    ) {
        self.account = account
        self.name = name
        self.address = address
        self.country = country
        self.identification = identification
    }
}

/// Financial institution information
public struct Institution: Codable {
    /// BIC (Bank Identifier Code)
    public let bic: String?
    
    /// Institution name
    public let name: String?
    
    /// Institution address
    public let address: String?
    
    /// Account number at the institution
    public let account: String?
    
    public init(
        bic: String? = nil,
        name: String? = nil,
        address: String? = nil,
        account: String? = nil
    ) {
        self.bic = bic
        self.name = name
        self.address = address
        self.account = account
    }
}

/// Charges indication for SWIFT messages
public enum ChargesIndication: String, Codable, CaseIterable {
    /// All charges for account of ordering customer
    case our = "OUR"
    
    /// All charges for account of beneficiary customer
    case ben = "BEN"
    
    /// Charges shared between ordering and beneficiary
    case sha = "SHA"
}

/// SWIFT message processing result
public struct SwiftProcessingResult: Codable {
    /// Transaction ID
    public let transactionId: String
    
    /// SWIFT reference
    public let swiftReference: String
    
    /// Processing status
    public let status: String
    
    /// Processing timestamp
    public let processingTime: Date
    
    /// Processing fees
    public let fees: Fee?
    
    /// Estimated settlement time
    public let estimatedSettlement: Date?
}

/// Fee information
public struct Fee: Codable {
    /// Fee amount
    public let amount: Decimal
    
    /// Fee currency
    public let currency: String
    
    /// Fee description
    public let description: String?
}

File: sdks/dotnet/Models/Swift/MT103Message.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace LegacyBaaS.SDK.Models.Swift
{
    /// <summary>
    /// MT103 Single Customer Credit Transfer message
    /// </summary>
    public class MT103Message
    {
        /// <summary>
        /// Gets or sets the transaction reference number
        /// </summary>
        [Required]
        [JsonPropertyName("transactionReference")]
        public string TransactionReference { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the value date
        /// </summary>
        [JsonPropertyName("valueDate")]
        public DateTime? ValueDate { get; set; }

        /// <summary>
        /// Gets or sets the currency code (ISO 4217)
        /// </summary>
        [Required]
        [StringLength(3, MinimumLength = 3)]
        [JsonPropertyName("currency")]
        public string Currency { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the transaction amount
        /// </summary>
        [Required]
        [Range(0.01, double.MaxValue)]
        [JsonPropertyName("amount")]
        public decimal Amount { get; set; }

        /// <summary>
        /// Gets or sets the ordering customer information
        /// </summary>
        [Required]
        [JsonPropertyName("orderingCustomer")]
        public Customer OrderingCustomer { get; set; } = new();

        /// <summary>
        /// Gets or sets the beneficiary customer information
        /// </summary>
        [Required]
        [JsonPropertyName("beneficiaryCustomer")]
        public Customer BeneficiaryCustomer { get; set; } = new();

        /// <summary>
        /// Gets or sets the ordering institution
        /// </summary>
        [JsonPropertyName("orderingInstitution")]
        public Institution? OrderingInstitution { get; set; }

        /// <summary>
        /// Gets or sets the beneficiary institution
        /// </summary>
        [JsonPropertyName("beneficiaryInstitution")]
        public Institution? BeneficiaryInstitution { get; set; }

        /// <summary>
        /// Gets or sets the intermediary institution
        /// </summary>
        [JsonPropertyName("intermediaryInstitution")]
        public Institution? IntermediaryInstitution { get; set; }

        /// <summary>
        /// Gets or sets the remittance information
        /// </summary>
        [StringLength(4 * 35)] // Max 4 lines of 35 characters
        [JsonPropertyName("remittanceInformation")]
        public string? RemittanceInformation { get; set; }

        /// <summary>
        /// Gets or sets the regulatory reporting information
        /// </summary>
        [JsonPropertyName("regulatoryReporting")]
        public string? RegulatoryReporting { get; set; }

        /// <summary>
        /// Gets or sets the charges indication
        /// </summary>
        [JsonPropertyName("charges")]
        public ChargesIndication? Charges { get; set; }
    }

    /// <summary>
    /// Customer information for SWIFT messages
    /// </summary>
    public class Customer
    {
        /// <summary>
        /// Gets or sets the account number
        /// </summary>
        [Required]
        [JsonPropertyName("account")]
        public string Account { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the customer name
        /// </summary>
        [Required]
        [StringLength(4 * 35)] // Max 4 lines of 35 characters
        [JsonPropertyName("name")]
        public string Name { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the customer address
        /// </summary>
        [StringLength(4 * 35)] // Max 4 lines of 35 characters
        [JsonPropertyName("address")]
        public string? Address { get; set; }

        /// <summary>
        /// Gets or sets the country code
        /// </summary>
        [StringLength(2, MinimumLength = 2)]
        [JsonPropertyName("country")]
        public string? Country { get; set; }

        /// <summary>
        /// Gets or sets additional identification
        /// </summary>
        [JsonPropertyName("identification")]
        public string? Identification { get; set; }
    }

    /// <summary>
    /// Financial institution information
    /// </summary>
    public class Institution
    {
        /// <summary>
        /// Gets or sets the BIC (Bank Identifier Code)
        /// </summary>
        [StringLength(11, MinimumLength = 8)]
        [JsonPropertyName("bic")]
        public string? Bic { get; set; }

        /// <summary>
        /// Gets or sets the institution name
        /// </summary>
        [StringLength(4 * 35)] // Max 4 lines of 35 characters
        [JsonPropertyName("name")]
        public string? Name { get; set; }

        /// <summary>
        /// Gets or sets the institution address
        /// </summary>
        [StringLength(4 * 35)] // Max 4 lines of 35 characters
        [JsonPropertyName("address")]
        public string? Address { get; set; }

        /// <summary>
        /// Gets or sets the account number at the institution
        /// </summary>
        [JsonPropertyName("account")]
        public string? Account { get; set; }
    }

    /// <summary>
    /// Charges indication for SWIFT messages
    /// </summary>
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum ChargesIndication
    {
        /// <summary>
        /// All charges for account of ordering customer
        /// </summary>
        OUR,

        /// <summary>
        /// All charges for account of beneficiary customer
        /// </summary>
        BEN,

        /// <summary>
        /// Charges shared between ordering and beneficiary
        /// </summary>
        SHA
    }

    /// <summary>
    /// SWIFT message processing result
    /// </summary>
    public class SwiftProcessingResult
    {
        /// <summary>
        /// Gets or sets the transaction ID
        /// </summary>
        [JsonPropertyName("transactionId")]
        public string TransactionId { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the SWIFT reference
        /// </summary>
        [JsonPropertyName("swiftReference")]
        public string SwiftReference { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the processing status
        /// </summary>
        [JsonPropertyName("status")]
        public string Status { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the processing timestamp
        /// </summary>
        [JsonPropertyName("processingTime")]
        public DateTime ProcessingTime { get; set; }

        /// <summary>
        /// Gets or sets any processing fees
        /// </summary>
        [JsonPropertyName("fees")]
        public Fee? Fees { get; set; }

        /// <summary>
        /// Gets or sets estimated settlement time
        /// </summary>
        [JsonPropertyName("estimatedSettlement")]
        public DateTime? EstimatedSettlement { get; set; }
    }

    /// <summary>
    /// Fee information
    /// </summary>
    public class Fee
    {
        /// <summary>
        /// Gets or sets the fee amount
        /// </summary>
        [JsonPropertyName("amount")]
        public decimal Amount { get; set; }

        /// <summary>
        /// Gets or sets the fee currency
        /// </summary>
        [JsonPropertyName("currency")]
        public string Currency { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the fee description
        /// </summary>
        [JsonPropertyName("description")]
        public string? Description { get; set; }
    }
}